src/utils/instrumentation.ts

```diff
@@ -0,0 +1,201 @@
+ * @param libraryName {string} The name of the logging library to ve reported. Should be prefixed with 'nodejs'.
+ *        Will be truncated if longer than 14 characters.
+ * @param libraryVersion {string} The version of the logging library to be reported. Will be truncated if longer than 14 characters.
+ * @returns {Entry} The entry with diagnostic instrumentation data.
+ */
+export function createDiagnosticEntry(
+  libraryName: string | undefined,
+  libraryVersion: string | undefined
+): Entry {
+  // Validate the libraryName first and make sure it starts with 'nodejs' prefix.
+  if (!libraryName || !libraryName.startsWith(NODEJS_LIBRARY_NAME_PREFIX)) {
+    libraryName = NODEJS_LIBRARY_NAME_PREFIX;
+  }
+  const entry = new Entry(
+    {
+      severity: google.logging.type.LogSeverity.INFO,
+    },
+    {
+      [DIAGNOSTIC_INFO_KEY]: {
+        [INSTRUMENTATION_SOURCE_KEY]: [
+          {
+            // Truncate libraryName and libraryVersion if more than 14 characters length
+            name: truncateValue(libraryName),
+            version: truncateValue(libraryVersion ?? getNodejsLibraryVersion()),
+          },
+        ],
+      },
+    }
+  );
+  return entry;
+}
+
+/**
+ * This method validates that provided instrumentation info list is valid and also adds current library info to a list.
+ * @param infoList {InstrumentationInfo} The array of InstrumentationInfo to be validated and updated.
+ * @returns {InstrumentationInfo} The updated list of InstrumentationInfo.
+ */
+function validateAndUpdateInstrumentation(
+  infoList: InstrumentationInfo[]
+): InstrumentationInfo[] {
+  const finalInfo: InstrumentationInfo[] = [];
+  // First, add current library information
+  finalInfo.push({
+    name: NODEJS_LIBRARY_NAME_PREFIX,
+    version: getNodejsLibraryVersion(),
+  });
+  // Iterate through given list of libraries and for each entry perform validations and transformations
+  // Limit amount of entries to be up to 3
+  let count = 1;
+  for (const info of infoList) {
+    if (isValidInfo(info)) {
+      finalInfo.push({
+        name: truncateValue(info.name),
+        version: truncateValue(info.version),
+      });
+    }
+    if (++count === 3) break;
+  }
+  return finalInfo;
+}
+
+/**
+ * A helper function to truncate a value (library name or version). The value is truncated
+ * when it is longet than 14 chars and '*' is added instead of truncated suffix.
+ * @param value {string} The library name or version to be truncated.
+ * @returns {string} The truncated value.
+ */
+function truncateValue(value: string) {
+  if (value && value.length > maxDiagnosticValueLen) {
+    return value.substring(0, maxDiagnosticValueLen).concat('*');
+  }
+  return value;
+}
+
+/**
+ * The helper function to retrieve current library version from 'package.json' file. Note that
+ * since we use {path.resolve}, the search for 'package.json' could be impacted by current working directory.
```

Comment:
What is the failure case for path.resolve and different current working directories?  Does path.resolve use the working directory of the process or the source code file as it appears here?

Reply:
Yes, it uses CWD as a base path. I had to make this change due to system tests defined in [install.ts](https://github.com/googleapis/nodejs-logging/blob/main/system-test/install.ts) - apparently those tests build a package and check if such package can be installed. Those tests were initially failing for me simply because I used relative path to retrieve a version, e.g. by running `require('../../package.json').version`.